ufs-community develop update from NCAR dtc/develop 2019/12/17

climbfuji opened the pull request:
This PR contains:
- update FMS to point to tag 2019.01 of NOAA-GFDL
- update stochastic_physics to point to hash [1745422af76d830757cd6035b6ea101e92b4cac1](https://github.com/noaa-psd/stochastic_physics/commit/1745422af76d830757cd6035b6ea101e92b4cac1) @pjpegion 
- update of path to CCPP physics library for CCPP regression tests
- add `-Wall` to compiler flags for GNU compilers
- remove warnings for non-existent include directories for GNU compiler
- update of regression testing scripts to detect errors in tests/run_test.sh
- new regression tests `fv3_ccpp_gfs_v15p2`, `fv3_ccpp_gfs_v15p2_debug`, `fv3_ccpp_gfs_v16beta`, `fv3_ccpp_gfs_v16beta_debug`
- update of `rt.conf`: remove Cheyenne.intel entries (so that `rt.conf` is an EMC-maintained regression test configuration; move Cheyenne Intel tests into a separate file that resides in the  NCAR dtc/develop branch)
- make `compile_cmake.sh` work with Cheyenne Intel/GNU, and bugfixes for `compile_cmake.sh`
- allow environment variable `NEMS_MACHINE` to overwrite (or set) `MACHINE_ID` (see https://github.com/ufs-community/ufs-weather-model/pull/20 for the corresponding PR for the ufs_public_release branch)
- new regression test `fv3_ccpp_gsd_sar_25km_debug` (but not exercised in default `rt.conf`)
- updates for jet and gaea (note: only supporting `rt.sh`, not `NEMSCompsetRun`)

Associated PRs:
https://github.com/NOAA-EMC/GFDL_atmos_cubed_sphere/pull/11
https://github.com/NCAR/ccpp-framework/pull/246
https://github.com/NCAR/ccpp-physics/pull/372
https://github.com/NOAA-EMC/fv3atm/pull/30
https://github.com/ufs-community/ufs-weather-model/pull/18

For regression testing, see this PR further down.

climbfuji commented:
Still todo:
- update this PR and associated PRs with the changes in the current commit (reference?).
- squash the commits into one commit (only ufs-weather-model; fv3atm done)

climbfuji commented:
Pre-testing on macOS with gcc+gfortran and with clang+gfortran: all build options work (IPD gmake, IPD cmake, CCPP dynamic gmake, CCPP static gmake, CCPP static cmake), and the fv3_gfdlmp/fv3_ccpp_gfdlmp regression tests give b4b identical results between all builds with one and two OpenMP threads.

climbfuji commented:
Pre-testing on Cheyenne with Intel using rt_ccpp_dtc.conf (create baseline using IPR REPRO, verify using CCPP dynamic REPRO and CCPP static REPRO): all tests pass.

[rt_ccpp_dtc_verify.log](https://github.com/ufs-community/ufs-weather-model/files/3991512/rt_ccpp_dtc_verify.log)
[rt_ccpp_dtc_create.log](https://github.com/ufs-community/ufs-weather-model/files/3991513/rt_ccpp_dtc_create.log)

climbfuji commented:
Pre-testing on Cheyenne with gcc+gfortran: all tests pass against existing baseline (maintained by DTC), except the two newly added tests for GFSv15p2 and GFSv16beta.

[rt_gnu.log](https://github.com/ufs-community/ufs-weather-model/files/3991517/rt_gnu.log)

climbfuji commented:
Pre-testing on Cheyenne with intel: all tests pass against existing baseline (maintained by DTC), except the two newly added tests for GFSv15p2 and GFSv16beta.

[rt_intel.log](https://github.com/ufs-community/ufs-weather-model/files/3993554/rt_intel.log)

climbfuji commented:
Pre-testing on hera with intel using rt.conf against official EMC baseline. The following tests fails:
```
bash-4.2$ grep -e FAIL log_hera.intel/*
log_hera.intel/rt_008_fv3_gfdlmprad_noahmp.log:Test 008 fv3_gfdlmprad_noahmp FAIL
log_hera.intel/rt_050_fv3_gfdlmprad_noahmp_repro.log:Test 050 fv3_gfdlmprad_noahmp FAIL
log_hera.intel/rt_055_fv3_gfdlmprad_repro.log:Test 055 fv3_gfdlmprad FAIL
log_hera.intel/rt_081_fv3_ccpp_gfdlmprad_repro.log:Test 081 fv3_ccpp_gfdlmprad FAIL
log_hera.intel/rt_093_fv3_ccpp_gfdlmprad_noahmp_repro.log:Test 093 fv3_ccpp_gfdlmprad_noahmp FAIL
log_hera.intel/rt_101_fv3_ccpp_gsd_repro.log:Test 101 fv3_ccpp_gsd FAIL
log_hera.intel/rt_113_fv3_ccpp_gfdlmprad_prod.log:Test 113 fv3_ccpp_gfdlmprad FAIL
log_hera.intel/rt_114_fv3_ccpp_wrtGauss_nemsio_c768_prod.log:Test 114 fv3_ccpp_wrtGauss_nemsio_c768 FAIL
log_hera.intel/rt_127_fv3_ccpp_gfdlmprad_noahmp_prod.log:Test 127 fv3_ccpp_gfdlmprad_noahmp FAIL
log_hera.intel/rt_135_fv3_ccpp_gsd_prod.log:Test 135 fv3_ccpp_gsd FAIL
log_hera.intel/rt_136_fv3_ccpp_gfs_v15p2_prod.log:Test 136 fv3_ccpp_gfs_v15p2 FAIL
log_hera.intel/rt_137_fv3_ccpp_gfs_v16beta_prod.log:Test 137 fv3_ccpp_gfs_v16beta FAIL
```

**All of these are expected failures, because:**
- NoahMP precipitation bugfix for both IPD and CCPP changes the answer of the following tests:
```
log_hera.intel/rt_008_fv3_gfdlmprad_noahmp.log:Test 008 fv3_gfdlmprad_noahmp FAIL
log_hera.intel/rt_050_fv3_gfdlmprad_noahmp_repro.log:Test 050 fv3_gfdlmprad_noahmp FAIL
log_hera.intel/rt_093_fv3_ccpp_gfdlmprad_noahmp_repro.log:Test 093 fv3_ccpp_gfdlmprad_noahmp FAIL
log_hera.intel/rt_127_fv3_ccpp_gfdlmprad_noahmp_prod.log:Test 127 fv3_ccpp_gfdlmprad_noahmp FAIL
```
- Addition of WW3 component to the following tests to match what EMC did for the IPD PROD version changes their results:
```
log_hera.intel/rt_055_fv3_gfdlmprad_repro.log:Test 055 fv3_gfdlmprad FAIL
log_hera.intel/rt_081_fv3_ccpp_gfdlmprad_repro.log:Test 081 fv3_ccpp_gfdlmprad FAIL
log_hera.intel/rt_113_fv3_ccpp_gfdlmprad_prod.log:Test 113 fv3_ccpp_gfdlmprad FAIL
log_hera.intel/rt_114_fv3_ccpp_wrtGauss_nemsio_c768_prod.log:Test 114 fv3_ccpp_wrtGauss_nemsio_c768 FAIL
```
- Updates of the GSD suite change the results of the following tests
```
log_hera.intel/rt_101_fv3_ccpp_gsd_repro.log:Test 101 fv3_ccpp_gsd FAIL
log_hera.intel/rt_135_fv3_ccpp_gsd_prod.log:Test 135 fv3_ccpp_gsd FAIL
```
- Newly added tests (as for UFS public release) do not have a baseline yet:
```
log_hera.intel/rt_136_fv3_ccpp_gfs_v15p2_prod.log:Test 136 fv3_ccpp_gfs_v15p2 FAIL
log_hera.intel/rt_137_fv3_ccpp_gfs_v16beta_prod.log:Test 137 fv3_ccpp_gfs_v16beta FAIL
```

[rt_against_original_emc_baseline_trunk-20191127.log](https://github.com/ufs-community/ufs-weather-model/files/4000844/rt_against_original_emc_baseline_trunk-20191127.log)

climbfuji commented:
@junwang-noaa @DusanJovic-NOAA @RatkoVasic-NOAA please have a look at this PR and the PRs listed in the description. I cleaned up the commit history, right now I am running the tests against the develop-20191217 baseline on hera.intel, and if everything is ok this will result in the same regression test failures as described above. Thanks! Dom

climbfuji commented:
Testing against current baseline (develop-20191217) after merging latest updates from EMC and cleaning up the commit history. On hera.intel, the same tests fail as described above - ok.

[rt_against_original_emc_baseline_trunk-20191217.log](https://github.com/ufs-community/ufs-weather-model/files/4007177/rt_against_original_emc_baseline_trunk-20191217.log)
[fail_test_rt_against_original_emc_baseline_trunk-20191217.log](https://github.com/ufs-community/ufs-weather-model/files/4007178/fail_test_rt_against_original_emc_baseline_trunk-20191217.log)

climbfuji commented:
Testing against current baseline (develop-20191217) after merging latest updates from EMC and cleaning up the commit history. On hera.intel using NEMSCompsetRun, test test `fv3_gfdlmprad_noahmp` fails as expected, all others pass.

[NEMSCompsetRun_against_original_emc_baseline_trunk-20191217.log](https://github.com/ufs-community/ufs-weather-model/files/4007481/NEMSCompsetRun_against_original_emc_baseline_trunk-20191217.log)
[rtreport_against_original_emc_baseline_trunk-20191217.txt](https://github.com/ufs-community/ufs-weather-model/files/4007482/rtreport_against_original_emc_baseline_trunk-20191217.txt)

climbfuji commented:
Testing on jet.intel: create new baseline (no old baseline exists), verify against it: all tests pass. The regression test logs (Compile_jet.intel.log, RegressionTests_jet.intel.log) have been updated in the PR.

[rt_create.log](https://github.com/ufs-community/ufs-weather-model/files/4008884/rt_create.log)
[rt_verify.log](https://github.com/ufs-community/ufs-weather-model/files/4008885/rt_verify.log)

climbfuji commented:
Testing on hera.intel: create new baseline, verify against it: all tests pass. The regression test logs (Compile_hera.intel.log, RegressionTests_hera.intel.log) have been updated in the PR.

[rt_create.log](https://github.com/ufs-community/ufs-weather-model/files/4010026/rt_create.log)
[rt_verify.log](https://github.com/ufs-community/ufs-weather-model/files/4010027/rt_verify.log)

climbfuji commented:
Testing on gaea.intel using Intel 18: create new baseline, verify against it: all tests pass. The regression test logs (Compile_gaea.intel.log, RegressionTests_gaea.intel.log) have been updated in the PR.

[rt_intel18_create.log](https://github.com/ufs-community/ufs-weather-model/files/4010030/rt_intel18_create.log)
[rt_intel18_verify.log](https://github.com/ufs-community/ufs-weather-model/files/4010031/rt_intel18_verify.log)

climbfuji commented:
Extra credit: I created a local version of `rt_gaea_intel16.conf` (see attachment) and modified `compile_cmake.sh` locally so that it sets the correct flags for CCPP and uses the correct modulefile when `INTEL16=Y` is found in the `MAKEOPT` string. I then created a baseline and verified against it (as usual, create for IPD PROD, IPD REPRO, CCPP PROD; verify for IPD PROD, IPD REPRO, CCPP REPRO, CCPP PROD). All tests pass.

The temporary baseline on gaea for Intel 16 can be found in `/lustre/f2/pdata/esrl/gsd/ufs/ufs-weather-model/RT/NEMSfv3gfs/develop-20191230_TEMPORARY_INTEL16/`

[rt_gaea_intel16_create.log](https://github.com/ufs-community/ufs-weather-model/files/4013064/rt_gaea_intel16_create.log)
[rt_gaea_intel16_verify.log](https://github.com/ufs-community/ufs-weather-model/files/4013065/rt_gaea_intel16_verify.log)
[rt_gaea_intel16.conf.txt](https://github.com/ufs-community/ufs-weather-model/files/4013066/rt_gaea_intel16.conf.txt)

The logs for Intel 16 are not committed to the repository, because there is no mechanism to support different versions of the same compiler for one machine, and because the default version on Gaea is Intel 18.

junwang-noaa commented on .gitmodules:
```diff
@@ -1,15 +1,15 @@
-	url = https://github.com/NOAA-EMC/FMS
-	branch = GFS-FMS
+	url = https://github.com/NOAA-GFDL/FMS
+	branch = master
```
Is the master same as the tag: 2019.01? Currently the tag 2019.01 has revision 708b8d5, while master has revision:  cb85999

DusanJovic-NOAA replied:
Git does not checkout HEAD of the master, but commit specified explicitly in the repository tree with command 'git add FMS'.

climbfuji replied:
Yes, I was just typing this: The PR points to tag 2019.01 as required, not to the head of master. There is no way to specify in .gitmodules that a certain tag is used, only branches can be specified. The tag 2019.01 was cut from master (I looked through the git log history in FMS), hence using master (or deleting the branch = line entirely, because master is the default) is correct imo.

junwang-noaa commented on tests/rt.sh:
```diff
@@ -159,7 +159,7 @@ elif [[ $MACHINE_ID = gaea.* ]]; then
   # DH* 20190717 temporary
   #DISKNM=/lustre/f2/pdata/ncep_shared/emc.nemspara/RT
-  DISKNM=/lustre/f2/pdata/esrl/gsd/gmtb/NEMSfv3gfs/RT
+  DISKNM=/lustre/f2/pdata/esrl/gsd/ufs/ufs-weather-model/RT
```
How was the gaea baseline created? Is it same as cray? I am copying RT baseline from cray to gaea as requested by GFDL. If this is done, I can let GFDL know.

climbfuji replied:
This baseline was created on gaea using Intel 18. As far as I understand, wcoss_cray is still using Intel 16. Once we are using the same version of the compiler, we should check if the baselines are indeed identical and start copying from wcoss_cray if yes.

junwang-noaa replied:
OK, all model baseline under emc.nemspara on gaea was copied from cray. Now we need to copy baseline from cray to gaea first before we can verify that the current code can still generate same regression test baseline on gaea. Would you please add an option such as "GSD_BASELINE" so that we can still use the emc.nemspara under ncep_shared for wcoss cray baseline? We can combine the ENC baseline with GSD-BASELINE the verification is done.

climbfuji replied:
You mean I should use something like this:
```
DISKNM=/lustre/f2/pdata/ncep_shared/emc.nemspara/RT
#DISKNM_GSD=/lustre/f2/pdata/esrl/gsd/ufs/ufs-weather-model/RT
```
But I don't understand how the baseline from cray should work for gaea if the compiler is different? Or are you using the exact same Intel version on cray? Gaea is 18.0.3.222

In my opinion we cannot make this requested change until we are using the same version of the compiler.

climbfuji replied:
Note also that on cray we are not running rt.conf but rt_wcoss_cray.conf, which only runs IPD PROD. This unification of baselines is something for a future commit. Here I am just re-enabling the regression tests on gaea and at the same time showing that CCPP and IPD are b4b identical even with they cray compilers.

junwang-noaa commented on tests/run_test.sh:
```diff
@@ -36,8 +36,7 @@ export REGRESSIONTEST_LOG=${LOG_DIR}/rt_${TEST_NR}_${TEST_NAME}${RT_SUFFIX}.log
 trap 'echo "run_test.sh: Test ${TEST_NAME} killed"; kill $(jobs -p); wait; trap 0; exit' 1 2 3 4 5 6 7 8 10 12 13 15
-./${RUN_SCRIPT} > ${RUNDIR_ROOT}/${TEST_NAME}${RT_SUFFIX}.log 2>&1 &
-wait
+./${RUN_SCRIPT} > ${RUNDIR_ROOT}/${TEST_NAME}${RT_SUFFIX}.log 2>&1
```
It's good to have this fixed! We also noticed that RT does not catch the error when generating new baseline. I think this fix will catch that error too.  Does this change impact the parallel run of RT tests?

climbfuji replied:
I will have to rely on Dusan to answer this question.

DusanJovic-NOAA replied:
I was using parallel rt.sh (using ecFlow) to create baselines and verify and I didn't have any issue with this change.

junwang-noaa commented on tests/tests/fv3_ccpp_sas:
```diff
@@ -0,0 +1,71 @@
+###############################################################################
+#
+#  FV3 CCPP SAS test
```
What is the tested based? GFSv15.2?

climbfuji replied:
It uses the ccpp_gfdlmp.nml.IN namelist template, i.e. it is basically the fv3_ccpp_gfdlmp test with SAS instead of SAMF (both shallow and deep). Same for the IPD version.

junwang-noaa commented on tests/tests/fv3_ccpp_satmedmfq:
```diff
@@ -0,0 +1,70 @@
+###############################################################################
+#
+#  FV3 CCPP satmedmfq test
```
what is base test?

climbfuji replied:
fv3_satmedmf / fv3_ccpp_satmedmf (i.e. the existing satmedmf test that uses the other/older version of the scheme)

junwang-noaa commented:
GFDL is working on this. They made a request for us to copy the cray
baseline to gaea. Originally the rt.sh was working on gaea. Later on, we
had NEMSCompsetrun working, so we were running NEMSCompsetrun on gaea
against cray baseline. I don't think there is anything wrong with ccpp
against IPD. We just want to confirm that the model can still generate same
results on gaea and cray with IPD, then we can switch to full RT test with
both IPD and CCPP.

On Tue, Dec 31, 2019 at 12:22 PM Dom Heinzeller <notifications@github.com>
wrote:

> *@climbfuji* commented on this pull request.
> ------------------------------
>
> In tests/rt.sh
> <https://github.com/ufs-community/ufs-weather-model/pull/18#discussion_r362251805>
> :
>
> > @@ -159,7 +159,7 @@ elif [[ $MACHINE_ID = gaea.* ]]; then
>    ECFLOW_START=
>    # DH* 20190717 temporary
>    #DISKNM=/lustre/f2/pdata/ncep_shared/emc.nemspara/RT
> -  DISKNM=/lustre/f2/pdata/esrl/gsd/gmtb/NEMSfv3gfs/RT
> +  DISKNM=/lustre/f2/pdata/esrl/gsd/ufs/ufs-weather-model/RT
>
> Note also that on cray we are not running rt.conf but rt_wcoss_cray.conf,
> which only runs IPD PROD. This unification of baselines is something for a
> future commit. Here I am just re-enabling the regression tests on gaea and
> at the same time showing that CCPP and IPD are b4b identical even with they
> cray compilers.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ufs-community/ufs-weather-model/pull/18?email_source=notifications&email_token=AI7D6TIKPHAC4R3UHNYLBETQ3N5VJA5CNFSM4J4M6I3KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQOJR2Y#discussion_r362251805>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI7D6TOMQH6BH7RFLYMM2QLQ3N5VJANCNFSM4J4M6I3A>
> .
>

climbfuji commented:
Ok, so in this case I recommend to update/revert DISKNM in a follow-up commit when GFDL has finished their work. It doesn't make sense to commit code that wasn't used for running the regression tests. The current commit clearly documents how the tests were run on gaea and what the results are.

> On Dec 31, 2019, at 10:29 AM, junwang-noaa <notifications@github.com> wrote:
> 
> GFDL is working on this. They made a request for us to copy the cray
> baseline to gaea. Originally the rt.sh was working on gaea. Later on, we
> had NEMSCompsetrun working, so we were running NEMSCompsetrun on gaea
> against cray baseline. I don't think there is anything wrong with ccpp
> against IPD. We just want to confirm that the model can still generate same
> results on gaea and cray with IPD, then we can switch to full RT test with
> both IPD and CCPP.
> 
> On Tue, Dec 31, 2019 at 12:22 PM Dom Heinzeller <notifications@github.com>
> wrote:
> 
> > *@climbfuji* commented on this pull request.
> > ------------------------------
> >
> > In tests/rt.sh
> > <https://github.com/ufs-community/ufs-weather-model/pull/18#discussion_r362251805>
> > :
> >
> > > @@ -159,7 +159,7 @@ elif [[ $MACHINE_ID = gaea.* ]]; then
> > ECFLOW_START=
> > # DH* 20190717 temporary
> > #DISKNM=/lustre/f2/pdata/ncep_shared/emc.nemspara/RT
> > - DISKNM=/lustre/f2/pdata/esrl/gsd/gmtb/NEMSfv3gfs/RT
> > + DISKNM=/lustre/f2/pdata/esrl/gsd/ufs/ufs-weather-model/RT
> >
> > Note also that on cray we are not running rt.conf but rt_wcoss_cray.conf,
> > which only runs IPD PROD. This unification of baselines is something for a
> > future commit. Here I am just re-enabling the regression tests on gaea and
> > at the same time showing that CCPP and IPD are b4b identical even with they
> > cray compilers.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/ufs-community/ufs-weather-model/pull/18?email_source=notifications&email_token=AI7D6TIKPHAC4R3UHNYLBETQ3N5VJA5CNFSM4J4M6I3KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQOJR2Y#discussion_r362251805>,
> > or unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/AI7D6TOMQH6BH7RFLYMM2QLQ3N5VJANCNFSM4J4M6I3A>
> > .
> >
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/ufs-community/ufs-weather-model/pull/18?email_source=notifications&email_token=AB5C2RK3WV6662H45FPHUGTQ3N6QLA5CNFSM4J4M6I3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH4PEEI#issuecomment-569963025>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB5C2RIYO5IOAGGTQWBWDRDQ3N6QLANCNFSM4J4M6I3A>.
>

climbfuji commented:
@DusanJovic-NOAA @junwang-noaa I reverted the change for fv3atm in .gitmodules and updated the submodule pointer following the merge of https://github.com/NOAA-EMC/fv3atm/pull/30. I also looked through all the changes again and couldn't find anything wrong (from my side).
